File: kuhn_cfr/node.py
from constants import *

class Node:
    def __init__(self, information_set: str):
        self.information_set = information_set
        self.cumulative_strategy = [0 for _ in range(NUM_ACTIONS)]
        self.cumulative_regrets = [0 for _ in range(NUM_ACTIONS)]
    
    def get_strategy(self):
        normalizing_sum = sum(filter(lambda x: x > 0, self.cumulative_regrets))
        if normalizing_sum > 0:
            return list(map((lambda x: x / normalizing_sum if x > 0 else 0), self.cumulative_regrets))
        else:
            return  [1 / NUM_ACTIONS for _ in range(NUM_ACTIONS)]

    def get_average_strategy(self):
        normalizing_sum = sum(self.cumulative_strategy)
        if normalizing_sum > 0:
            return list(map((lambda x: x / normalizing_sum), self.cumulative_strategy))
        else:
            return [1 / NUM_ACTIONS for _ in range(NUM_ACTIONS)]
    
    def update_average_strategy(self, strategy, realization_weigth: float):
        for a in range(NUM_ACTIONS):
            self.cumulative_strategy[a] += strategy[a] * realization_weigth
    
    def accumulate_action_regret(self, action_index, regret):
        self.cumulative_regrets[action_index] += regret

    def __str__(self):
        out = f"{self.information_set}:\t"
        average_strategy = self.get_average_strategy()
        
        for a in range(NUM_ACTIONS):
            out += f"{ACTIONS[a]}: {(average_strategy[a] * 100):.2f} %\t"
        return out

File: kuhn_cfr_cpp/node.cpp
#include "node.h"
#include "constants.h"
#include <sstream>
#include <iomanip>

Node::Node(const std::string &p_informationSet) : m_informationSet(p_informationSet), m_cumulativeStrategy{0, 0}, m_cumulativeRegrets{0, 0}
{
}
double *Node::GetStrategy() const
{
    double normalizingSum = 0;
    for (std::size_t a = 0; a < NUM_ACTIONS; a++)
    {
        if (m_cumulativeRegrets[a] > 0)
        {
            normalizingSum += m_cumulativeRegrets[a];
        }
    }
    double *strategy = new double[NUM_ACTIONS];
    if (normalizingSum > 0)
    {
        for (std::size_t a = 0; a < NUM_ACTIONS; a++)
        {
            strategy[a] = m_cumulativeRegrets[a] > 0 ? m_cumulativeRegrets[a] / (double)normalizingSum : 0;
        }
    }
    else
    {
        for (std::size_t a = 0; a < NUM_ACTIONS; a++)
        {
            strategy[a] = 1 / (double)NUM_ACTIONS;
        }
    }
    return strategy;
}
double *Node::GetAverageStrategy() const
{
    double normalizingSum = 0;
    for (std::size_t a = 0; a < NUM_ACTIONS; a++)
    {
        normalizingSum += m_cumulativeStrategy[a];
    }
    double *averageStrategy = new double[NUM_ACTIONS];
    if (normalizingSum > 0)
    {
        for (std::size_t a = 0; a < NUM_ACTIONS; a++)
        {
            averageStrategy[a] = m_cumulativeStrategy[a] / (double)normalizingSum;
        }
    }
    else
    {
        for (std::size_t a = 0; a < NUM_ACTIONS; a++)
        {
            averageStrategy[a] = 1 / (double)NUM_ACTIONS;
        }
    }
    return averageStrategy;
}
void Node::UpdateCumulativeStrategy(const double *p_strategy, double p_realizationWeight)
{
    for (std::size_t a = 0; a < NUM_ACTIONS; a++)
    {
        m_cumulativeStrategy[a] += p_strategy[a] * p_realizationWeight;
    }
}
void Node::AccumulateActionRegrets(int p_actionIndex, double p_regret)
{
    m_cumulativeRegrets[p_actionIndex] += p_regret;
}
std::string Node::ToString() const
{
    std::ostringstream stream;
    double *averageStrategy = GetAverageStrategy();
    stream << m_informationSet << ":\t";
    for (std::size_t a = 0; a < NUM_ACTIONS; a++)
    {
        stream << ACTIONS[a] << ": " << std::fixed << std::setprecision(2) << averageStrategy[a] * 100 << " %\t";
    }
    delete[] averageStrategy;
    return stream.str();
}
